### ISSUE TITLE
App receives only first telemetry packet on connect, but then stops updating the values of sensors

### ISSUE THREAD

**Issue by DiscoMan18** [zosko/iNavTelemetry#4]
iPhone XS, FrSky QX7, BLE HM-10 module (not sure what speed it's set, but it works with this app https://github.com/CrazyDude1994/android-taranis-smartport-telemetry/releases receiving telemetry from QCZEK LRS)

Right after connect the app shows only first telemetry packet containing values of some sensors, like current, voltage, RSSI, but not all of them. GPS data is not present at all. And the app then completely stops updating the values of sensors.

**Comment by zosko** [comment]
@DiscoMan18 what protocol you use ? 

Can you try this beta version ? i prepare new version in next days with bug fixes.

https://testflight.apple.com/join/7jfCnnmV

**Comment by zosko** [comment]
@DiscoMan18  is **QCZEK LRS** support Smart Port ? try to download last version from AppStore and test it.

**Comment by DiscoMan18** [comment]
Joined the Testflight, downloaded and tried latest version of the app. The issue remained the same. I guess it happens due to a difference in serial port speed of the HM-10 and your application which requires 4800 bps. I don't know what speed the HM-10 is set to. As it's already working with android taranis sp telemetry app it's probably set to 57600 bps. I can't change its speed cause it's a part of my working setup that I use on everyday basis. Could you add a dialog or an option that allows changing serial port speed of chosen communication device?

**Comment by zosko** [comment]
Communication between HM-10 and Application is not related to Baud. But HM-10 is related with other side like FlightController or Other. Can you tell me more about connection ? 

Bluetooth just send data from device to application. Let me know what protocol you use and how you connect to bluetooth so i can help you.

**Comment by DiscoMan18** [comment]
Well, in more detail you can read about QLRS [here](qczek.beyondrc.com), but let me explain. On my quadcopter I use iNav 2.6.1 which sends Mavlink telemetry to the QLRS RX module. The telemetry feed is being received by ground relay that consists of QLRS TX and Frsky R-XSR receiver. QLRS TX module outputs S.PORT telemetry (converted to this format from received mavlink stream) and R-XSR is used to forward S.PORT telemetry to my Taranis QX7, which then shows all the received data on the screen. 

In the internal bay of the QX7 I have HM-10 BLE module connected to S.PORT output. Other apps can connect to this module to display telemetry, map etc. Android Taranis Smartport telemetry app (which I mentioned in the first post) connects successfully to this module and works great. I tried your app but had issues I described earlier.

**Comment by zosko** [comment]
Okey so you using QLRS and relay that to XSR receiver to your QX7 transmitter. In QX7 you are using like this one
https://github.com/iNavFlight/OpenTX-Telemetry-Widget. And because you have built in HM-10 that mean you not need inverter.

in iNavTelemetry iOS application` (last version 1.1.1)` you choose `S.Port` from menu and connect to your HM-10 bluetooth ? Is it connection good or you cant connect ? If connection is good and you not receive any telemetry data, that mean there is some wrong in telemetry. Let me know what happen in last version. If still not receive data i will make custom version for you so we can debug problem.

**Comment by DiscoMan18** [comment]
Yes, I do what you describe. The connection is good, the app connects to hm-10 module and receives, as I wrote earlier, only the first telemetry packet which contains data not from all sensors. Some data is not present. With the last version of the app I observe the same behaviour, nothing changed here.

**Comment by zosko** [comment]
@DiscoMan18  try last version from AppStore, now also available for MacOS. If still same issue, then seems you dont have S.Port protocol, maybe `QCZEK LRS` use Crossfire protocol. That still is not developed. Im planing to do next when i receive ExpressLRS.

**Comment by DiscoMan18** [comment]
Tried last version from Appstore and also OS X version. Both show the same behaviour. Only the first telemetry packet is received (or probably a few of first packets) but then all the sensors stop updating. 

QCZEK LRS use S.PORT protocol for sure, as I see received telemetry on the screen of my Taranis Q-X7 (using Lua telemetry script. You can read more about the data sent to the radio [here](https://qczek.beyondrc.com/qczek-lrs-433mhz-1w-lora-rc-link/qczek-lrs-sport-telemetry/) Please follow this link and check the sensor id's, are they the same that your app expects to see in the telemetry stream.

**Comment by zosko** [comment]
Can you tell me what is DISTANCE telemetry value if that support ? You can check that in this screen
![](http://qczek.beyondrc.com/wp-content/uploads/2018/10/qlrs_sport3.jpg)

but to compare with this this screen in bottom left side
![](http://qczek.beyondrc.com/wp-content/uploads/2018/10/qlrs_sport4.jpg)

meantime i will prepare version for testing

**Comment by DiscoMan18** [comment]
Distance telemetry value is not sent by QLRS. It's calculated in my Taranis (you can setup a special calculated sensor right in the settings) as a 2D distance between home point (first gps coordinate received) and next coordinates. 

I've sent you a couple of screenshots (via Testflight app) so you'll be able to see what sensors the app receives. I changed the heading and pitch/roll of the quadcopter by moving it by hand, then pressed "connect". The app received new position of the quad with no subsequent updates whatsoever. At this point I made first screenshot. The I pressed "disconnect", again by hand changed heading and pitch/roll of the quad, then pressed "connect". The app received and showed new values of corresponding sensors. At this point I made the second screenshot.

**Comment by zosko** [comment]
@DiscoMan18 i have added 2.0.2 (3) version, please test that https://testflight.apple.com/join/7jfCnnmV

See if packet counting, and also there is debug screen to view packets that are not processed. That you can see in debug screen.

To open Debug Screen, click 5 times on Packets Counts rectangle.

Let me know after this.

**Comment by DiscoMan18** [comment]
Ok, just tested new version... Received packets are counting, got to debug screen, made the screenshot, sent it to you. The telemetry sensors still not updating properly, no changes from previous release here.

**Comment by zosko** [comment]
Thanks for quick feedback, i separate and others packets, to see are you receiving any others. Now i present "valid / invalid" packets. I push new version for testing, let me know how it will be. Just send me screenshots from debug screen if appear some packet there. And let me know how much are valid and invalid received. Last version is 2.0.2 (4)

**Comment by DiscoMan18** [comment]
Not at all, you made a great looking and very promising app, so I'm very interested in having it fully functional and stable, that's why I'm glad to help! 

I've sent you 2 screenshots, the first one is from the main screen and the 2nd from the debug screen which is empty now. According to numbers, seems like that the app receives only 13 valid packets and all the packet arriving after are invalid. See it yourself.

**Comment by zosko** [comment]
Looks you are right, im not sure why QCZEK LRS sent only 13 packets valid. Just to know all packets are good, but i separate in valid / invalid. Valid packet are that values i show on screen, invalid values are other packets that i not show but still they are good :) .

However, im not sure why is sending only 13 packets, seems 13 packets are all fields presented like Distance, Altitude, Speed, Current... they are 13 :) .
Im not sure why is only one time send that value. Can you verify is it same S.Port works on Android app ? because how i see QCZEK LRS support and MavLink. Maybe i can code that protocol and you can try that. 

Let me know if is posible you to test on Android app same S.Port. Is it working better or its same.  

For now i can verify that connection is good between mobile and transmitter, but there is missing data. Maybe better is to switch on MavLink only telemetry and connect to app, i can do that next but will take some time, plus you will be my tester :)

**Comment by DiscoMan18** [comment]
Yes, S.PORT is working good on my radio and it produces valid telemetry values that Androis Taranis Smartport Telemetry App recognises successfully. This is the application I'm currently flying with. 

I can only guess, but from my point of view there is something wrong with the logic you use to separate received packets and consider them as valid or invalid. Seems like that the packets your app denies as being invalid are still good and contain useful data.

**Comment by zosko** [comment]
Seems QCZEK LRS sending MavLink also, maybe that protocol is detected, i have made some experimental version of MavLink V1 / V2 . Im not sure what version is working on QCZEK LRS. I can prepare version for you to test that version.

**Comment by DiscoMan18** [comment]
It would be nice to have mavlink telemetry too, so thanks in advance for that, but let me explain again the situation we have with QLRS and S.PORT protocol. 

Yes, QCZEK sends down to ground also Mavlink telemetry, but it's only available at TX (ground unit) module on a certain pin. However on the ground I use a relay (Frsky R.XSR + QCZEK TX) which is bidirectional in terms both of RC Link from ground to air and telemetry, which is converted by QCZEK TX to S.PORT protocol and forwarded to my Taranis by telemetry channel of R.XSR receiver. So I'm 100% sure that Taranis receives S.PORT protocol (as it can display received telemetry values on its own screen). In the same time S.PORT telemetry is available on the connector in internal bay. This is the point where HM-10 bluetooth is connected to.

**Comment by zosko** [comment]
@DiscoMan18 can you check this experimental version: 2.1.0 (1) ? its on testflight. Is it possible somehow to extract data that you transfer from transmitter to mobile ? I can parse that hex data and see what is going on. Or even better, is it possible to install that software on Arduino / TTGO ESP32 if i dont have that device ? so i can extract that data. Im saying this because i can find code from that software to see how they send it. Only what i can find is this empty repo
https://github.com/krzysztofkuczek/QCZEK_LRS

**Comment by DiscoMan18** [comment]
Will do. In the meantime I asked other guys from QLRS thread on RCGROUPS to try your app and report to this thread. More testers is always good!

**Comment by DiscoMan18** [comment]
Tried version 2.1.0. It's working much better! Some sensors finally started to get new values! Check the screenshots I sent to you. However the update rate is very bad. For example heading and pitch/roll are updated only once in 20-30 seconds. Voltage is updated more or less frequently, current draw still shows 0.

**Comment by RomanLut** [comment]
@DiscoMan18  You can send telemetry log file recorded by "Android Taranis SmartPort Telemetry Viewer" to @zosko. It is actually unmodified binary stream which is coming from bluetooth. Then @zosko could feed it to app and find out why packets are not parsed correctly.

**Comment by zosko** [comment]
@DiscoMan18  cool, so that means packets are not good, if you can grab telemetry log from android application so i can see what is received. I see from screenshots s.port is detected. But packets are wrong. I see in android app they support and other types from s.port i will try to implement that. Mean time if you can send the log.

**Comment by DiscoMan18** [comment]
Ok, here is the telemetry log I saved from Android app. It contain the whole flight data, including gps coordinates, current draw, voltage, satellite count, speed, altitude etc. Hope this helps. 

[2020-08-02 19-50-51.log.zip](https://github.com/zosko/iNavTelemetry/files/7391353/2020-08-02.19-50-51.log.zip)
.

**Comment by zosko** [comment]
@DiscoMan18 i tested `22382` packets. From that i did not process `12823` packets because i did not use in application, here bellow what type of package is not processed.

type: F000  << PGR_PGN_VERSION_MASK
type: 100  << APID_ALTITUDE ( i think is from barometar, but i use from GPS)
type: 110   << APID_VARIO 
type: A00  << APID_AIR_SPEED
type: 4A0  << APID_MAV_BASE_MODE
type: 4A1  << APID_MAV_SYS_STATUS
type: 4A2  << APID_MAV_CUSTOM_MODE
type: 4B0  << APID_CUST_RSSI
type: 4B1  << APID_RX_RSSI_REG_VAL
type: 4B2 << APID_RX_SNR_REG_VAL
type: 4B3  << APID_RX_PACKET_LOST_VAL
type: F104 << RX_BAT
type: F105 << UNKNOWN_PACKET_1
type: F106  << i dont know what packet is this

all other 13 types of sensors are processed ~700 times per value.  So from this i see all is good in application about parser. I will do some other version where i can see if i can improve something.

**Comment by DiscoMan18** [comment]
Glad to hear that! As the new version arrives, I'll test it asap. BTW, if you plan adding support for various LRS (Like ExpressLRS for example) it would be great to have a possibility to configure displayed sensors. Choose from several kind of RSSI, prefer GPS Altitude or Barometer altitude, of display both, etc. 

Regarding RSSI, in case of QCZEK LRS, two values make the most use: 4B1 which represents the strength of received signal in dBm and 4B2 which is SNR - signal to noise ratio in dB.

**Comment by zosko** [comment]
Hey @DiscoMan18 i put new version `2.1.0 (2)`. Please check that and let me know.

Regarding RSSI, im using only SIGNAL from S.Port because other values i can only show in RAW format, dont know how to calculate to show 0% to 100% signal strength.

**Comment by DiscoMan18** [comment]
Tested new version. The update rate of heading and attitude (pitch and roll) still is very slow, only once in 20-30 seconds. There is still no current draw shown. RSSI sensors are shown both and the update rate is rather good. 

Regarding RSSIs (and other sensors), if possible, please add an option to show raw values. And also the ability to choose map source. Openstreet map, google (satellite, terrain and hybrid) etc.

**Comment by DiscoMan18** [comment]
Any update my friend?

**Comment by zosko** [comment]
Hey @DiscoMan18 i was improving speed in parsing but i did not tested long time ago, i was super busy, i just upload that version on TestFlight. `2.1.0 (3)`  you can try it. I have tested on MSP and S.Port protocol, on mine side speed is good and fine, all packets received on time and looks good. Connection between iPhone and FrSky Taranis is over Bluetooth 4.0.

Meantime i was order ExpressLRS parts, so i made DIY version with TTGO_V2 (TX) and MiniRX. Soon i will make Crossfire protocol because ExpressLRS is working on that protocol.

**Comment by DiscoMan18** [comment]
Hi @zosko I just tested latest version. No changes of the update rate. Most sensors are updated only once in 7-12 seconds.
